### DISCUSSION TITLE
Building a library used in packaged and unpackaged apps

### DISCUSSION THREAD

**Discussion by mlynch** [microsoft/WindowsAppSDK#1868] [Q&A] [answered]
I have a library that depends on the Windows App SDK (1.0), and users might target both packaged and unpackaged apps with it. There are [many APIs](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-supported-api#apis-that-require-package-identity) that seem to require package identity and, thus, don't work in unpackaged apps. 

How can I, as a library author, defensively program around this? Is there a way to detect whether the app is packaged or unpackaged at runtime? What about replacements for some of these APIs like `FileIO` that are very useful? Any help for managing these types of unknowns at runtime would be very helpful.

**Comment by mveril** [comment, marked as answer]
Hello
This documentation page could be useful for your
[Detect package identity and runtime context](https://docs.microsoft.com/windows/msix/detect-package-identity)

**Reply by mlynch** [reply]
Thank you, that's very helpful!

Far and away the biggest APIs I'm using that I'm struggling to adapt between packaged and unpackaged are the `Windows.Storage` and `Windows.ApplicationModel.Resources` APIs. Do you know of alternatives to those I should look into?

**Reply by andrewleader** [reply]
For `ApplicationModel.Resources`, have you looked into using MRT Core which is part of the WinAppSDK and should (theoretically) work the same for packaged and unpackaged? https://docs.microsoft.com/windows/apps/windows-app-sdk/mrtcore/mrtcore-overview

For Storage/file access, you'll have to handle the differences yourself right now.

**Reply by mlynch** [reply]
I’ll take a look at that, thanks! Storage is the bigger challenge. Any hints as to what alternatives would be there? Or is the alternative using normal C# standard library file classes?

**Reply by andrewleader** [reply]
For unpackaged apps, storage is completely free-form and up to the app developer how they want to store files.

The best practice is to store app data within the %AppData% folder on the PC, you can get the file path to that folder from C# using this StackOverflow answer: https://stackoverflow.com/a/867493

Note that it gives you the generic AppData folder, and then you have to choose to use a unique folder for your app/service/library within there.

You can use the standard .NET File APIs to work with reading and writing files (like [System.IO.File](https://docs.microsoft.com/en-us/dotnet/api/system.io.file?view=net-6.0)).

Let me know if you have any other questions!

**Reply by mlynch** [reply]
Okay thanks that makes a lot of sense. Basically the managed `Storage` utilities are just doing a lot of that stuff for you in the context of this `package` system. So, outside of that, I need to emulate it. Shouldn't be too difficult then. Thanks!

**Reply by ptorr-msft** [reply]
A lot of the answer depends on the type of packaged app, and the locations of the files you're expecting to operate on. If all the files are part of the app (either in the app's package, or in the app's "private storage") then you can use any API you want. If the files are found in other places (photo library, documents folder, random user-provided filenames, etc.) then your options are little different for AppContainer apps (note that there's no concern for just general packaged apps). 

If possible, for a library that _might_ be used in AppContainer apps and _might_ have to deal with arbitrary file locations,  my recommendation would be to avoid filenames altogether and simply accept streams / strings / byte buffers (with encoding, if appropriate) instead. That can negatively affect the usability of your API, and you might have requirements to really have the filename for other reasons, but using streams or buffers as your currency avoids the permissions problems.

**Reply by DrusTheAxe** [reply]
>Is there a way to detect whether the app is packaged or unpackaged at runtime?
>...
> This documentation page could be useful for your [Detect package identity and runtime context](https://docs.microsoft.com/windows/msix/detect-package-identity)

You may find [AppModel::Identity::IsPackagedProcess()](https://github.com/microsoft/WindowsAppSDK/blob/main/dev/Common/AppModel.Identity.h) a better reference model<sup>1</sup> for how to detect if you're running in a packaged or unpackaged process (i.e. has/lacks package identity).

<sup>1</sup> I'm fairly confident making this statement as I wrote both those functions, plus the underlying API they use :-)

**Reply by andrewleader** [reply]
That AppModel API isn't public? Or it's not projected to C#? Or not part of 1.0? I'd love it if we had a simple API for this but I'm not seeing it in my C# app.

**Reply by DrusTheAxe** [reply]
That's not an API per se, just some code in the project using the recipe (but as noted, done well 🙂).

GetCurrentPackageFullName is a public Windows exported Flat-C API. There's no C# equivalent or projection, but a simple P/Invoke would do it.

It never crossed anyone's mind to make it an API given it's such a simple thing but now that you mention it, yes that would make sense as an API. Please create a new Issue requesting that functionality and we can follow up through that

Thanks for the ask/suggestion!
